cranelift/codegen/src/machinst/vcode.rs

```diff
@@ -1115,6 +1116,43 @@ impl<I: VCodeInst> VCode<I> {
         }
     }
 
+    fn monotonize_inst_offsets(&self, inst_offsets: &mut [CodeOffset], func_body_len: u32) {
+        if self.debug_value_labels.is_empty() {
+            return;
+        }
+
+        // During emission, branch removal can make offsets of instructions incorrect.
+        // Consider the following sequence: [insi][jmp0][jmp1][jmp2][insj]
+        // It will be recorded as (say):    [30]  [34]  [38]  [42]  [<would be 46>]
+        // When the jumps get removed we are left with (in "inst_offsets"):
+        // [insi][jmp0][jmp1][jmp2][insj][...]
+        // [30]  [34]  [38]  [42]  [34]
+        // Which violates the monotonicity invariant. This method sets offsets of these
+        // removed instructions such as to make them appear zero-sized:
+        // [insi][jmp0][jmp1][jmp2][insj][...]
+        // [30]  [34]  [34]  [34]  [34]
+        //
+        let mut next_offset = func_body_len;
+        for inst_index in (0..(inst_offsets.len() - 1)).rev() {
+            let inst_offset = inst_offsets[inst_index];
+            if inst_offset > next_offset {
+                trace!(
+                    "Fixing code offset of the removed Inst {}: {} -> {}",
+                    inst_index,
+                    inst_offset,
+                    next_offset
+                );
+                inst_offsets[inst_index] = next_offset;
+                continue;
+            }
+
+            // Not all instructions get their offsets recorded.
+            if inst_offset != 0 {
```

Comment:
This does seem to leave an interesting corner case: if the very first instruction is a jump and is chomped (say because it's a jump to the fallthrough location, i.e. the next block), we artificially skip.

However I think it's OK for the following very subtle reason: we can only chomp back to an empty buffer (and hence have an offset of `0` on the non-first instruction) if we have one jump instruction (and nothing else) in the first block; MachBuffer will never chomp e.g. a cond+uncond pair to nothing. In the single-jump-inst case, that inst also has an offset of `0` so it's still monotonic.

Could we leave a comment to this effect?

Reply:
That is a great point you bring up.

> if we have one jump instruction (and nothing else) in the first block; MachBuffer will never chomp e.g. a cond+uncond pair to nothing. In the single-jump-inst case, that inst also has an offset of 0 so it's still monotonic.

Am I right to understand that this is because of the VCode construction? I considered:
```
Block0:
  jmp Block1
  jmp Block1
Block1:

; Which _can_ be simplified down to nothing.
```
I was writing the comment as follows, does that match your expectation?
```
// Not all instructions get their offsets recorded. As for the case where
// the offset is legitimately zero (first instruction), we are guaranteed
// that no more than one prior jump (which would also have been at a zero
// offset) could have been removed because the first block will not have
// multiple jumps.
```

Reply:
I would replace the last bit with "... because either the first block will not have multiple branches, or else it would not have been completely elided."

This is related to your example: no block will have two unconditional branches; we emit only one "terminator" VCode instruction per block, and those are either one uncond, or a two-target form that is cond+uncond (or switch or ret, both of which are indirect).

(Additional subtlety just to allay any confusion: some pseudoinsts *do* emit branch insts internally, but those are completely invisible to the MachBuffer (i.e., not reported).)

Reply:
I've thought about this some more, and became less convinced it's a correct assumption. Consider:
```
Block0:
  jmp Block2
Block1:
  jmp Block2
Block2:
```
(This can be fixed in a more obvious matter by using not-zero for the distinguished "invalid offset" value)

Reply:
Ah! Yes, indeed, that's the far more sensible solution. (One might be tempted to make it a `Vec<Option<CodeOffset>>` but for memory-efficiency let's define a sentinel value e.g. `const NO_INST_OFFSET: CodeOffset = u32::MAX` maybe?)